processing-service-core/api/models/tenant.js

```diff
@@ -0,0 +1,340 @@
+
+var Q = require('q');
+var logger = require('../helpers/loggingHelper.js');
+var tenantConfigStoreModel = require("./db/tenantConfigStore.js");
+var globalConfigStoreModel = require("./db/globalConfigStore.js");
+var apiErrorTypes = require('./errors/apiErrorTypes.js');
+var ApiError = require('./errors/apiError.js');
+
+
+module.exports = {
+    setTenantConfig: setTenantConfig,
+    setTenantConfigs: setTenantConfigs,
+    deleteTenantConfig: deleteTenantConfig,
+    deleteTenantConfigs: deleteTenantConfigs,
+    getEffectiveTenantConfig: getEffectiveTenantConfig,
+    getEffectiveTenantConfigs: getEffectiveTenantConfigs,
+    getTenantConfig: getTenantConfig,
+    getTenantConfigs: getTenantConfigs
+};
+
+/**
+ * Creates or updates tenant specific config.
+ * Creates or updates a custom configuration setting for this key against this tenant if one doesn't exist or updates the currently 
+ * existing custom configuration setting if it does.
+ *
+ * @param {String} tenantId String The ID of the tenant.
+ * @param {String} key String The custom configuration key.
+ * @param {String} value String The value to use when setting the custom config for the tenants key. (optional)
+ * no response value expected for this operation
+ **/
+function setTenantConfig(tenantId, key, value) {
+    var tenantConfig = Q.defer();
+    logger.debug("Creating tenant specific config for tenant " + tenantId + " with key " + key + " and value " + value);
+    tenantConfigStoreModel.setTenantConfig(tenantId, key, value)
+            .then(function (createConfigPromise) {
+                logger.debug("Configuration of " + key + " set for tenant " + tenantId + " with value " + value);
+                tenantConfig.resolve();
+            })
+            .fail(function (errorResponse) {
+                logger.debug("Unable to set configuration for key " + key + " for tenant " + tenantId + " with value " + value);
+                tenantConfig.reject(errorResponse);
+            })
+            .done();
+    return tenantConfig.promise;
+}
+
+/**
+ * Creates or updates tenant specific configs.
+ * Creates or updates a custom configuration setting for this key against the tenant.
+ *
+ * @param {String} tenantId  String The ID of the tenant.
+ * @param {String} tenantConfigs TenantConfigs The json string representation key-value pair map of custom tenant configuration.
+ * no response value expected for this operation
+ **/
+function setTenantConfigs(tenantId, tenantConfigs) {
+    var tenantConfig = Q.defer();
+    var promiseArray = [];
+
+    for (var i = 0; i < tenantConfigs.configurations.length; i++) {
+        var config = tenantConfigs.configurations[i];
+        promiseArray.push(tenantConfigStoreModel.setTenantConfig(tenantId, config.key, config.value));
+    }
+    Q.all(promiseArray)
+            .then(function (results) {
+                logger.debug("Created tenant specific config for tenant " + tenantId);
+                tenantConfig.resolve();
```

Comment:
This could be a mix of creates and deletes. We could check if all the results returned are creates, otherwise consider this an update.

Reply:
Withdrawing comment. Decision is to always return 204